### ISSUE TITLE
Error running containers: Error loading seccomp filter into kernel

### ISSUE THREAD

**Issue by racaceres** [moby/moby#45498]
### Description

I have a server running some containers with docker-compose and the `restart: always` option, so when a container fails, Docker restarts it. The problem is since some months ago Docker shows this error when tried to start a container: 

```
OCI runtime exec failed: exec failed: unable to start container process: error loading seccomp filter into kernel: error loading seccomp filter: errno 524: unknown
```

I tried updating Docker, Docker compose, kernel and passing a custom seccomp configuration, but nothing of this was work. The only workarounds for the problem are to set the `seccomp:unconfined` or to restart the docker daemon, but it are not viable at the long term because one is a security error and the other causes downtime and the problem appears again in a few weeks.

Anyone can help me to figure it out how to fix this issue?

### Reproduce

1. Run some containers
2. Wait some weeks of normal operation (including restart containers)
3. Eventually container restarts will fails with that error

### Expected behavior

The kernel must load the seccomp filters without problems always.

### docker version

```bash
Client:
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.18.1
 Git commit:        20.10.21-0ubuntu1~22.04.2
 Built:             Thu Mar  2 18:26:04 2023
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.21
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.1
  Git commit:       20.10.21-0ubuntu1~22.04.2
  Built:            Wed Feb 15 15:26:57 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.12-0ubuntu1~22.04.1
  GitCommit:        
 runc:
  Version:          1.1.4-0ubuntu1~22.04.1
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc., v2.16.0)

Server:
 Containers: 86
  Running: 59
  Paused: 0
  Stopped: 27
 Images: 3535
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 
 runc version: 
 init version: 
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.0-60-generic
 Operating System: Ubuntu 22.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 40
 Total Memory: 62.8GiB
 Name: ---
 ID: YLDP:SGFK:UX5E:WM77:PM6P:NWXF:6LHM:F6R5:IH4U:34P3:EONM:7S7I
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

```
$ uname -r
5.15.0-60-generic
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.2 LTS"
```

**Comment by akerouanton** [comment]
Hi @racaceres, the same issue has been reported to EKS:

- https://github.com/awslabs/amazon-eks-ami/issues/1179#issuecomment-1429765381

Can you try to increase the value of `net.core.bpf_jit_limit` as the linked comment says?

**Comment by thaJeztah** [comment]
@akerouanton curious now if that's the cause, if that's something we need to document? (or even recommend changing / automatically change on install?) /cc @corhere @tianon

**Comment by corhere** [comment]
Document maybe, [recommend no.](https://bugs.launchpad.net/ubuntu/+source/libseccomp/+bug/1956954/comments/7)

If it is, in fact, the BPF JIT resource limit being hit, I find it quite curious that restarting _the docker daemon_ successfully works around the issue. And reading between the lines, it sounds like there is a non-trivial amount of container churn happening. Unless one of the workloads is loading huge (number of) BPF programs into the kernel, it sounds a lot like the JIT'ed BPF programs for seccomp filters—or perhaps cgroup eBPF programs?—are not being cleaned up when a container exits, but are being cleaned up when dockerd (or containerd?) does. My hunch is that bumping `net.core.bpf_jit_limit` is working around a resource leak and that the root cause is a bug in runC, systemd or the kernel.

**Comment by racaceres** [comment]
Thanks for the useful responses, I check the `net.core.bpf_jit_limit` on my server and its minor than the suggested on the linked post. 

```
~ $ sudo sysctl --all | grep net.core.bpf_jit_limit
net.core.bpf_jit_limit = 264241152
~ $ sudo cat /proc/vmallocinfo | grep bpf_jit | awk '{s+=$2} END {print s}'
91500544
```

Unfortunately, yesterday we restarted the docker daemon (`systemctl restart docker`), so right know the problem its not happening, but when it happens again we'll increase the `bpf_jit_limit` value and check if it solves the problem. 

I'll close this issue for the moment and I'll update it when the solution has been checked.

Thanks a lot!

**Comment by akerouanton** [comment]
@corhere Indeed, this commit torvalds/linux@3a15fb6ed92cb32b0a83f406aa4a96f28c9adbc3 introduced a memory leak in seccomp (first released in v5.9). It was fixed in torvalds/linux@a1140cb215fa13dcec06d12ba0c3ee105633b7c4 and has been released in v6.2. Here's the list of backports:

- v6.1.2 (torvalds/linux@5b81f0c6c60e35bf8153230ddfb03ebb14e17986)
- v6.0.16 (torvalds/linux@29a69fa075d0577eff1137426669de21187ec182)
- v5.15.86 (torvalds/linux@a31a647a3d1073a642c5bbe3457731fb353cb980)

About the default value of `net.core.bpf_jit_limit`, this patch is doubling it: torvalds/linux@10ec8ca8ec1a2f04c4ed90897225231c58c124a7. It's been released in v6.3 and backported to:

- v6.2.9 (torvalds/linux@68ed00a37d2d1c932ff7be40be4b90c4bec48c56)
- v6.1.22 (torvalds/linux@9cda812c76067c8a771eae43bb6943481cc7effc)
- v5.15.105 (torvalds/linux@54869daa6a437887614274f65298ba44a3fac63a)
- v5.10.177 (torvalds/linux@a4bbab27c4bf69486f5846d44134eb31c37e9b22)
- v5.4.240 (torvalds/linux@d69c2ded95b17d51cc6632c7848cbd476381ecd6)
- v4.19.280 (torvalds/linux@42049e65d338870e93732b0b80c6c41faf6aa781)
- v4.14.312 (torvalds/linux@374ed036309fce73f9db04c3054018a71912d46b)

As such, I don't think we should document `bpf_jit_limit`.

@racaceres As said by @corhere increasing `bpf_jit_limit` is only a workaround. You should probably update your kernel.

**Comment by racaceres** [comment]
@akerouanton Thanks for such detailed explaination, with it I've a better understanding of the root cause of the problem. 

At the moment, the last available kernel version on the Ubuntu 22.04 LTS repos is the `5.15.0.71.69` but I'll be attentive to future updates.

**Comment by akerouanton** [comment]
Looks like this fix was shipped in Ubuntu's `5.15.0-68.75` kernel, so you should be all good with the latest kernel available :wink:

**Comment by racaceres** [comment]
Great! thanks so much.

**Comment by racaceres** [comment]
I updated my server with Ubuntu 22.04LTS to the kernel `5.15.0-71-generic` and increase the value of `net.core.bpf_jit_limit` to `452534528` but the problem appears again. In that moment the amount of memory used by bpf_jit was `1044123648` more than 2x the limit. 

The only way I can think to solve the problem is to upgrade my server to a non-LTS version of ubuntu with a newest kernel, but Perhaps is not a good idea for a production server. Is there any other way to solve the problem that doesn't involve using a non-LTS version?

**Comment by thaJeztah** [comment]
If it's an issue with Ubuntu / Canonical's kernel, it might be worth checking if they back ported the fix; https://kernel.ubuntu.com/git/ubuntu/, and otherwise opening a ticket in their issue tracker; https://bugs.launchpad.net/ubuntu/+source/linux/+filebug

**Comment by Hrishikesh143** [comment]
We encountered an error during pod creation where the container runtime failed with a seccomp initialization error (unable to init seccomp: error loading seccomp filter into kernel). After investigation, we identified that libseccomp was not installed on the affected machine. Since runc depends on libseccomp to load and apply seccomp profiles during container startup, the absence of this library caused the container initialization to fail. Installing libseccomp on the node using **$sudo apt install libseccomp** and rebooting the machine resolved the issue, and pods started successfully afterward.
